File: src/me/Salt/Container/CommandParser.java
/*
 * Copyright (c) 2017 DevJake
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.Salt.Container;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Project title: SaltBot-2.0
 * Created by Salt on 10/04/2017.
 * <p>
 * This class parses raw text into a range of details about a command
 */


public class CommandParser {
    public CommandContainer parse(String raw, String cmdPrefix) {
        String beheaded = raw.replaceFirst(cmdPrefix, "");
        List<String> splitBeheaded = Arrays.asList(beheaded.split(" "));
        List<String> argsUpper = splitBeheaded.subList(1, splitBeheaded.size());
        List<String> argsLower = new ArrayList<>();
        argsUpper.forEach(s -> argsLower.add(s.toLowerCase()));
        String cmd = splitBeheaded.get(0).toLowerCase();
        String cmdUpper = splitBeheaded.get(0);

        return new CommandContainer(raw, argsLower, argsUpper, beheaded, splitBeheaded, cmd, cmdUpper);
    }


    public class CommandContainer {
        private String rawText;
        private List<String> argsLower;
        private List<String> argsUpper;
        private String beheaded;
        private List<String> splitBeheaded;
        private String cmd;
        private String cmdUpper;

        /**
         * @param rawText       String - The raw text of the command
         * @param argsLower     List - A list of the arguments, in lowercase
         * @param argsUpper     List - A list of the arguments, in the original case they were initially entered in
         * @param beheaded      String - The raw text, with the command prefix removed
         * @param splitBeheaded List - The raw text processed through a beheading (removal of the command prefix), then separated by a space
         * @param cmd           String - The command itself, in lowercase
         * @param cmdUpper      String - The command itself, in the original case it was initially entered in
         */
        public CommandContainer(String rawText, List<String> argsLower, List<String> argsUpper, String beheaded, List<String> splitBeheaded, String cmd, String cmdUpper) {
            this.rawText = rawText;
            this.argsLower = argsLower;
            this.argsUpper = argsUpper;
            this.beheaded = beheaded;
            this.splitBeheaded = splitBeheaded;
            this.cmd = cmd;
            this.cmdUpper = cmdUpper;
        }

        @Override
        public String toString() {
            return "CommandContainer{" +
                    "rawText='" + rawText + '\'' +
                    ", argsLower=" + argsLower +
                    ", argsUpper=" + argsUpper +
                    ", beheaded='" + beheaded + '\'' +
                    ", splitBeheaded=" + splitBeheaded +
                    ", cmd='" + cmd + '\'' +
                    ", cmdUpper='" + cmdUpper + '\'' +
                    '}';
        }

        public String getRawText() {
            return rawText;
        }

        public List<String> getArgsLower() {
            return argsLower;
        }

        public List<String> getArgsUpper() {
            return argsUpper;
        }

        public String getBeheaded() {
            return beheaded;
        }

        public List<String> getSplitBeheaded() {
            return splitBeheaded;
        }

        public String getCmd() {
            return cmd;
        }

        public String getCmdUpper() {
            return cmdUpper;
        }
    }
}

File: src/main/kotlin/me/salt/entities/cmd/CommandParser.kt
/*
 * Copyright 2017 DevJake
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.salt.entities.cmd

import me.salt.entities.config.Configs
import me.salt.entities.config.entities.GuildConfig
import me.salt.entities.config.entities.SaltConfig
import me.salt.entities.config.entities.TextChannelConfig
import me.salt.entities.config.entities.UserConfig
import me.salt.entities.objects.Entity
import me.salt.entities.objects.exists
import me.salt.entities.objects.getConfig
import me.salt.util.events.CommandParseEvent
import me.salt.util.events.fireEvent
import me.salt.util.exception.*
import me.salt.util.logging.logDebug
import java.util.regex.Pattern

object CommandParser {
    fun isPotentialCommand(raw: String) = Pattern.compile("^[^a-zA-Z]+").matcher(raw.split(" ")[0]).find()

    fun parse(raw: String, guildId: String, textChannelId: String, userId: String): CommandContainer {
        var saltMain: SaltConfig? = null
        var guildMain: GuildConfig? = null
        var textMain: TextChannelConfig? = null
        var userMain: UserConfig? = null

        if (Configs.salt.MAIN_CONFIG.exists())
            saltMain = Configs.salt.MAIN_CONFIG.getConfig(SaltConfig::class.java)

        if (Configs.guild(guildId).MAIN_CONFIG.exists())
            guildMain = Configs.guild(guildId).MAIN_CONFIG.getConfig(GuildConfig::class.java)

        if (Configs.textChannel(guildId).MAIN_CONFIG.exists())
            textMain = Configs.textChannel(textChannelId).MAIN_CONFIG.getConfig(TextChannelConfig::class.java)

        if (Configs.user(guildId).MAIN_CONFIG.exists())
            userMain = Configs.user(userId).MAIN_CONFIG.getConfig(UserConfig::class.java)

        val beheaded = mutableMapOf<Entity, Pair<String, String>>()

        /*
        Determines which prefixes are applicable, and filters them into the beheaded variable.
        This allows for specific information to be distributed in events, for example.
         */
        saltMain?.globalPrefixes?.forEach {
            if (raw.startsWith(it))
                beheaded.put(Entity.GLOBAL, Pair(it, raw.replaceFirst(it, "")))
        }

        guildMain?.guildPrefixes?.forEach {
            if (raw.startsWith(it))
                beheaded.put(Entity.GUILD, Pair(it, raw.replaceFirst(it, "")))
        }

        textMain?.channelPrefixes?.forEach {
            if (raw.startsWith(it))
                beheaded.put(Entity.TEXTCHANNEL, Pair(it, raw.replaceFirst(it, "")))
        }

        userMain?.preferredCmdPrefixes?.forEach {
            if (raw.startsWith(it))
                beheaded.put(Entity.USER, Pair(it, raw.replaceFirst(it, "")))
        }

        var beheadedLiteral: String = ""

        if (!Pattern.compile("^[^a-zA-Z]+").matcher(raw.split(" ").get(0)).find())
            exception(PrefixlessCommandException("The given input, $raw, seems to lack any sort of prefix!"))
        if (!Pattern.compile("^.*?[^a-zA-Z]+").matcher(raw.split(" ").get(0)).find())
            exception(AmorphousCommandException("The given input, $raw, has a foul-structured (non-identifiable) prefix!"))
        else
            beheadedLiteral = raw.replace(Regex("^.*?[^a-zA-Z]+"), "")
        if (beheaded.isEmpty())
            exception(AmorphousCommandException("The given input, $raw, failed to match any prefixes loaded from configs!"))

        //TODO If the raw is identified as matching to a command, recursively remove the first character until reaching a letter
        //No specific way to check and appropriately remove identified prefixes without looping over them each removal; inefficient and needless
        //TODO https://github.com/DevJake/SaltBot-2.0/blob/master/src/main/java/me/Salt/Command/Container/CommandParser.java
        //TODO ensure prefixes cannot **end** with a letter (a-z)

        val cc = CommandContainer(
                beheaded,
                beheadedLiteral,
                beheadedLiteral.toLowerCase(),
                beheadedLiteral.split(" "),
                beheadedLiteral.split(" ").map { it.toLowerCase() }
        )

        fireEvent(CommandParseEvent(cc))
        logDebug("Parsed command; CommandContainer=$cc")
        return cc
    }

    data class CommandContainer(
            val beheaded: MutableMap<Entity, Pair<String, String>>,
            val beheadedLiteral: String,
            val beheadedLiteralLower: String,
            val args: List<String>,
            val argsLower: List<String>)


}
